DOCS-5467-Access-Management-Connected-Apps (#48)
mulesoft/docs-access-management: modules/ROOT/pages/connected-apps-permissions.adoc

= Connected Apps Permissions by User
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> Connected Apps can be used by Organization Administrators, developers, and end-users; each of these roles has different privileges when connecting applications,

Reviewer: ```suggestion
Connected apps can be used by organization administrators, developers, and end users; each of these roles has different privileges when connecting applications,
```
I'm not sure about organization administrators. That is an actual role so it might need to remain title case and be bolded.  Kim will know.

and there are some restrictions to ensure sensitive data remains secure.

== Organization Administrator

> An Organization Administrator is the only role that can view and manage Connected Apps in Access Management.

Reviewer: I'm not sure about capping Connected Apps. I would lc all, but Kim might have diff advice.

Author: ```suggestion
An Organization Administrator is the only role that can view and manage connected apps in Access Management.
```

Author: I don't think it should be capped unless I'm referring to the page at this point.

> An Organization Administrator is the only role that can view and manage Connected Apps in Access Management.

Reviewer 2: ```suggestion
Organization administrators are the only users that can view and manage connected apps in Access Management.
```

To manage your connected apps, navigate to Access Management and select the *Applications* tab.

// screenshot of Connected Apps page

> The *Connected Apps* page shows a list of applications that you have connected to Anypoint Platform.

Reviewer: ```suggestion
The *Connected Apps* page shows a list of apps that are connected to Anypoint Platform.
```

Reviewer: Technically, this is a procedure and should be an ol.

To add an application, select *Add Application*.

> === Create Application

Reviewer: Create an App

> === Create Application

Author: ```suggestion
=== Create an App
```

> When you create an application as an Organization Administrator, consider the following:

Reviewer: ```suggestion
When you create an app as an Organization Administrator, consider the below:
```

Author: ```suggestion
When you create an app as an organization administrator, consider the below:
```

* An organization can own up to 200 connected apps.

> * You can disable or whitelist connected apps.

Reviewer 2: Whitelisting connected apps is orthogonal to their creation.

We have two personas:
**A developer** that wants to register and administrate an Application that they've made. 
This includes:
- setting grant types, client names, redirect uris
- Enabling/Disabling/Deleting the client if its secret has been leaked
- Viewing usage about their client (How many organizations are using it? How many users?)

**An Administrator** that wants to control how their organization's data is exposed.
This includes:
- Viewing which users have shared their data to which apps, and revoking share access if necessary
- Maintaining the list of apps that users are allowed to share data to (The whitelist)
- Maintaining settings about how data can be shared
    - What grant types can be used? Is the whitelist enabled?

Today, both of those personas are performed by Organization Administrators. But, I'd be very deliberate about separating the two when talking about them.

A more concrete example: Github
I own one Github app. To see it, I navigate to Developer Settings
![image](https://user-images.githubusercontent.com/11035210/63310888-3b377180-c2b1-11e9-8387-06fc6a1c27e0.png)

I've also authorized that Github app to access my data. To see that, I navigate to my Personal Settings
![image](https://user-images.githubusercontent.com/11035210/63310931-5c985d80-c2b1-11e9-9f19-7512f0612c0e.png)

I am both a producer of my app and a consumer. But, my app is still treated like everyone else's. The majority of our users will only be app consumers, not creators.

Author: I'm going to follow up with you on explaining this a little more carefully to users - I am still pretty unsure of where devs fit into this equation.

Reviewer 3: @sarathecoubian this is the most important piece of feedback. There's org admins creating and managing applications on one side, and there's users connecting applications, and their org admins managing their authorizations and whitelists on the other. The two are very distinct and I believe they need separate pages.

> * An application is assigned a clientID upon creation; the clientID cannot be modified. However, client secrets can be modified in the *Application Settings* page.

Reviewer: ```suggestion
* When you create an app, a clientID is assigned, which can't be modified. However, you can modify client secrets in the *Application Settings* page.
```

> * When you create an application with the client grant type, the application will continue to work if the application owner's credentials are disabled for any reason. A team or company can assume control of the application's access.

Reviewer: ```suggestion
* When you create an application with the OAuth client credential grant type, the application continues to work even if the application owner's credentials are disabled for any reason. A team or company can assume control of the application's access.
```

> * When you create an application with the client grant type, the application will continue to work if the application owner's credentials are disabled for any reason. A team or company can assume control of the application's access.

Reviewer 2: There is no such thing as an application owner for Connected Apps. ConnApps are owned by an organization, and any Org Admin in the administration can control the app.

> * When you create an application with the client grant type, the application will continue to work if the application owner's credentials are disabled for any reason. A team or company can assume control of the application's access.

Author: ```suggestion
* When you create an application with the OAuth client credential grant type, the app is owned by the organization; any organization administrator can control the app.
```

> * An application's actions are logged with an application ID and user ID (for user grant type applications only) in the Audit Log. This empowers administrators to monitor and audit data use.

Reviewer: Please verify the grant type. AFAIK, there is not a "user" grant type for OAuth. It would be a **password** grant type, where the user name and password are passed.

Reviewer 2: We've been using the terminology User Grant Type to represent the grant types that require user interaction ('implicit', 'authorization_code', 'refresh_token') and Client Grant Type to represent the grant types that do not ('client_credentials').

> * An application's actions are logged with an application ID and user ID (for user grant type applications only) in the Audit Log. This empowers administrators to monitor and audit data use.

Reviewer: ```suggestion
* For apps that use the password grant type, actions taken by the app are logged with an application ID and user ID in the audit log so administrators can monitor and audit data use.
```

Reviewer 2: I'd avoid mentioning password grant type - that's a separate thing from client_credentials that we haven't implemented for Connected Apps (And don't plan on doing so either!)

Author: ```suggestion
* For apps that use the user grant type, actions taken by the app are logged with an application ID and user ID in the audit log so administrators can monitor and audit data use.
```

// screenshot of Create Application interface

> When creating an application, you must enter the following information:

Reviewer: ```suggestion
When creating an application, you must enter the below information:
```

> * *Application Name*: A descriptive name for your application.

Reviewer: Kim wants us to do list items like this:

- Item
   Definition
- Item 
   Definition

Basically, like a DL

> * *Application Name*: A descriptive name for your application.

Author: ```suggestion
* *Application Name*:: 
A descriptive name for your application.
```

> * *Redirect URIs*: Uniform Resource Identifier that specifies the location of the application. You can specify one per line.

Reviewer 2: ```suggestion
* *Redirect URIs*::
 Uniform Resource Identifier that specifies the location of the application's callback. You can specify one per line.
```

> * *User Grant Types* or *Client Grant Types*: A grant type that aligns with OAuth 2.0 specifications.

Reviewer: This looks like UI, so there must be a user grant type. But please do verify what that means because, again, in the world of OAuth 2.0 I don't know of a "user" grant type. Not that I'm an expert.

Reviewer: Just verify that if it is indeed a "user" grant type, does that mean the user name and password are passed, or...?

Reviewer 2: Mentioned above - we are not supporting the password grant type. We've been using user grant types as a term to mean Grant Types that require user input

> * *User Grant Types* or *Client Grant Types*: A grant type that aligns with OAuth 2.0 specifications.

Reviewer 2: ```suggestion
* *Allowed Grant Types*: A list of OAuth 2.0 grant types the application intends to make use of
```

> It is important to specify the correct grant type when connecting an application since it will dictate who can access the application. Note that client grant types are only available for first-party applications.

Reviewer 2: It's the other way around - who can access the application dictates what grant types are allowed. If an application is made public, client grant types are not allowed.
First party applications, if made public, can't have the client grant type - since if they are made public they are someone else's third party application

> It is important to specify the correct grant type when connecting an application since it will dictate who can access the application. Note that client grant types are only available for first-party applications.

Author: ```suggestion
It is important to specify the access permissions when connecting an app since that will dictate the available grant types. Note that client grant types are only available for private first-party apps. 
```

==== User Grants

> User grants require authentication from a user, and they are primarily useful for third-party applications. Connected Apps currently supports two types of user grant authentication:

Reviewer: Authentication meaning user name and password?

> User grants require authentication from a user, and they are primarily useful for third-party applications. Connected Apps currently supports two types of user grant authentication:

Reviewer 2: ```suggestion
Clients making use of User grant types require authorization from a user. Connected Apps currently supports two methods of receiving authorization.
```

Author: ```suggestion
Clients making use of user grant types require authorization from a user. Connected Apps currently supports two methods of receiving authorization.
```

> * Authorization: requires the application to produce an authorization code that is accepted by OAuth 2.0.

Reviewer: This is bugging me...I guess they are grouping diff grant types under "user." Strange. Because authorization code is an actual grant type.  I'm not commenting on this anymore.  :-)

> * Authorization: requires the application to produce an authorization code that is accepted by OAuth 2.0.

Reviewer 2: ```suggestion
* Browser-based Authorization Flow::
 The app directs the user's browser (commonly called a User-Agent) to Anypoint, where the user authorizes the app before being redirected back
```

> If the code should not work indefinitely, you can select the *Refresh Tokens* option, requiring new codes as often as specified in the application.

Reviewer: ```suggestion
Select the *Refresh Tokens* option to require new codes as often as specified in the application rather than allowing code to work indefinitely.
```

> If the code should not work indefinitely, you can select the *Refresh Tokens* option, requiring new codes as often as specified in the application.

Reviewer 2: - Codes must be exchanged for access tokens
- they can only be exchanged once
- access tokens ALWAYS expire within a short period of time, usually a matter of hours, or when the user logs out

If the app requires access after the user logs out, or doesn't want to prompt them every time, they can use refresh tokens.
- The app must have the authorization_code grant type to use the refresh_token grant type
- Codes _may_ be exchanged for an (access token, refresh token) pair iff the user approves it
- refresh tokens may be exchanged by the app at a later date for another (access token, refresh token) pair

Author: How much does a user already know about refresh tokens?

> * JTW Bearer: grants access using an Application Public Key in a Privacy Enhanced Mail (PEM) that validates the JWT Authorization.

Reviewer: ```suggestion
* JTW Bearer
  Grants access using an application public key in a Privacy-Enhanced Mail (PEM) that validates 
  the JWT Authorization.
```

Reviewer 2: JWT Bearer apps are private apps that the Organization Administrator registers with a public key. The App then uses the corresponding private key to sign JWTs that may then be exchanged for an access token. When an Organization Administrator registers the app, they authorize it to retrieve tokens for all users in the org. Because of this, it is still considered a user grant type, even though no users will go through the authorization flow. In addition, individual users cannot revoke access to their account.
I wouldn't mention PEM files - we just have private and public keys.

> * JTW Bearer: grants access using an Application Public Key in a Privacy Enhanced Mail (PEM) that validates the JWT Authorization.

Author: ```suggestion
* JTW Bearer:: 
A private app that the organization administrator registers with a public key. The app uses the corresponding private key to sign JWTs that may then be exchanged for an access token. 
```

> After you select one or more user grant types, you can specify the scope of permissions.

Reviewer: ```suggestion
You can specify the scope of permissions after you select one or more user grant types.
```

> After you select one or more user grant types, you can specify the scope of permissions.

Reviewer 2: Scope of permissions is a confusing turn of phrase.
For apps using a user grant type, you specify the list of scopes the app is allowed to request from a user. Each scope represents a set of permissions, identical to how a role represents a set of permissions. Unlike roles, you do not assign the scopes with context, since the context comes from the user who authorizes the app.

For apps using a client grant type, you assign scopes with a context. These assignments can be thought of as identical to assigning an app a role directly. (And they are! just in a OAuth 2.0 flavored / spec compliant way)

> After you select one or more user grant types, you can specify the scope of permissions.

Author: ```suggestion
You can specify the list of scopes the app is allowed to request from a user after you select one or more user grant types.
```

==== Client Grants

> Client grants are only permitted for first-party applications; they give users access based on their organizational affiliations.

Reviewer: ```suggestion
Client grants give users access based on their organizational affiliations and are permitted only for first-party apps.
```

> You can also create more granular access restrictions by specifying which environments have access.

Reviewer 2: Another weird turn of phrase - you specify which environments the app may access - not which environments may access the app

> You can also create more granular access restrictions by specifying which environments have access.

Author: ```suggestion
You can also specify which environments the app can access.
```

> After you select an organization or organization and environment, you can specify the scope of permissions.

Author: ```suggestion
You can specify the list of scopes the app is allowed to request from a user after you select an organization or organization and environment.
```

== Developer

> Developers who have websites or web apps can allow users to log in using the *Log in with Anypoint Platform* option.

Reviewer: I'm not sure web sites is one word.

Reviewer: Also, this is a UI thing, but they should change that UI label to Sign in instead of Log in. The nomenclature on the Anypoint Platform sign-in page is sign in.

Author: I shall bring it up. From what I can tell, websites is a single word, but I imagine Kim will have an opinion.

> For this feature to work, the connected application must be configured with user grant authentication.

Reviewer 2: We're still fleshing out the details on what Login with Anypoint entails, but there's a few more things coming 
The one I can guarantee is that the app must be configured with the `openid` scope at bare minimum

Author: I will follow up with you on this.

// code/chunk of steps for websites/web apps

> == End-user

Author: ```suggestion
== End User
```

> An end-user can log into a connected application using their Single Sign-On credentials. End-users can authorize access when the following message appears:

Reviewer: I don't know if end-user is hyphenated. I think it's just end user, but Kim will know for sure.

> An end-user can log into a connected application using their Single Sign-On credentials. End-users can authorize access when the following message appears:

Reviewer 2: Another weird phrase for me - end users authenticate themselves to Anypoint using whatever means they have configured, whether that's SSO or regular username/password. Once authenticated to Anypoint they can authorize any apps they want.

Reviewer 2: Connected Applications also allows Anypoint to be the SSO provider for these apps. So there's 1-2 levels of SSO, ex.
`Okta -> Anypoint -> My App` when external identity is configured for the user's org
`Anypoint -> My App` when the user is logging in with a username and password 
I'm not sure how to capture intent here

Author: Not really sure how or where to capture the second half of your comment; I am reorganizing this document's information, more than likely, so I will follow up with you to figure out where this should go.

> An end-user can log into a connected application using their Single Sign-On credentials. End-users can authorize access when the following message appears:

Author: ```suggestion
An end user can log into a connected application using their Anypoint credentials. End users can also  authorize access when the following message appears:
```

> // screenshot of authorize user page https://qax.anypoint.mulesoft.com/login/authorize?state=eyJjbGllbnQiOnsiY2xpZW50X2lkIjoiZWEzYzJiODFkMDEyNGI0Mjk1NDYyM2RhMWM0MzQ5ZWIifSwidXNlciI6eyJ1c2VybmFtZSI6Im9yZ2FuaXphdGlvbl9vd25lciJ9LCJvcmlnaW4iOiIvYXBpL3YyL29hdXRoMi9hdXRob3JpemU%2FY2xpZW50X2lkPWVhM2MyYjgxZDAxMjRiNDI5NTQ2MjNkYTFjNDM0OWViJnJlZGlyZWN0X3VyaT1odHRwczovL2V4YW1wbGUuY29tJnJlc3BvbnNlX3R5cGU9dG9rZW4mc2NvcGU9ZnVsbCIsInRyYW5zYWN0aW9uX2lkIjoiSXlNVGVJSHciLCJzY29wZSI6WyJmdWxsIl19

Reviewer 2: The Authorize page is changing soon @LeoWW let's make sure we keep the docs in sync with UI changes

Reviewer 4: Ack -- I'll update Sara when UI features are pushed

Reviewer 4: FYI: Here's the mockup for what the final Connected Apps UI will look like: https://www.figma.com/file/nGfK0OXqbSk15WoLN4dO2Q/Connected-apps?node-id=0%3A1

> End-users can manage third-party applications to which they have granted access in their profile settings.

Reviewer: ```suggestion
End-users can manage third-party applications to which they have access in their profile settings.
```

Should it be **Profile** settings? Is it UI?

= Connected Apps Permissions by User
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> Connected Apps can be used by organization administrators, developers, and end users; each of these roles has different privileges when connecting applications,

Reviewer 4: Nit: This feels like a run on sentence. I suggest using a period :) 

```
Connected Apps can be used by organization administrators, developers, and end users**.** Each of these.....
```

> Each of these roles has different privileges when connecting applications 

What is meant by **connecting applications**? This sounds a little vague.

Author: ```suggestion
Connected Apps can be used by organization administrators, developers, and end users.
```

> Connected Apps can be used by organization administrators, developers, and end users; each of these roles has different privileges when connecting applications,

Reviewer 3: please standardize on one term for connecting an application: Using / connecting / authorizing.

> and there are some restrictions to ensure sensitive data remains secure.

Reviewer 3: What restrictions? what sensitive data?

== Organization Administrator

Organization administrators are the only users that can view and manage connected apps in Access Management.

> To manage your connected apps, navigate to Access Management and select the *Applications* tab.

Reviewer 4: Note: I think the **Applications** tab will be renamed to **Connected Apps** in the release

// screenshot of Connected Apps page

The *Connected Apps* page shows a list of apps that are connected to Anypoint Platform.
To add an application, select *Add Application*.

=== Create an App

When you create an app as an organization administrator, consider the below:

* An organization can own up to 200 connected apps.
* You can disable or whitelist connected apps.

> * When you create an app, a clientID is assigned, which can't be modified. However, you can modify client secrets in the *Application Settings* page.

Reviewer 4: suggestion: a _unique_ clientId is assigned

Author: ```suggestion
* When you create an app, a unique clientID is assigned; the clientID can't be modified. However, you can modify client secrets in the *Application Settings* page.
```

> * When you create an application with the OAuth client credential grant type, the app is owned by the organization; any organization administrator can control the app.

Reviewer 3: This is true of any grant type, not just client_credentials

> * When you create an application with the OAuth client credential grant type, the app is owned by the organization; any organization administrator can control the app.

Reviewer 3: client_credentials

> * For apps that use the user grant type, actions taken by the app are logged with an application ID and user ID in the audit log so administrators can monitor and audit data use.

Reviewer 3: there's no "user grant type". There are several grant types that an application can use to act on behalf of a user: JWT Bearer, authorization_code, and basically every grant type that's not client_credentials

// screenshot of Create Application interface

When creating an application, you must enter the below information:

* *Application Name*:

 A descriptive name for your application.

* *Redirect URIs*:

>  Uniform Resource Identifier that specifies the location of the application's callback. You can specify one per line.

Reviewer 3: What's an application callback? Let's explain it further

* *Allowed Grant Types*: 

 A list of OAuth 2.0 grant types the application intends to make use of. It is important to specify the access permissions when connecting an app since that will dictate the available grant types. Note that client grant types are only available for private, first-party apps. 

==== User Grants

Clients making use of user grant types require authorization from a user. Connected Apps currently supports two methods of receiving authorization.

* Browser-based Authorization Flow:

 The app directs the user's browser (commonly called a User-Agent) to Anypoint, where the user authorizes the app before being redirected back. If the code should not work indefinitely, you can select the *Refresh Tokens* option, requiring new codes as often as specified in the application.

> * JTW Bearer:

Reviewer 4: I think it's JWT Bearer

Author: ```suggestion
* JWT Bearer:
```

Author: It definitely is - whoops. Good catch.
